### ISSUE TITLE
Defining multiple resources always returns first resource of the given request method

### ISSUE THREAD

**Issue by marcbrevoort-cyberhive** [viniciusgerevini/http-test-server#10]
TL;DR: Found a bug and raised a pull request (#12) to fix it.

Thank you for providing a useful testing tool. Running on `http-test-server = { version = "2.1.0" }`

When defining a http_test_server object with more than one canned response, this always seems to return the first one, rather than the response matching the resource URI.

An external test case for this might look something like
```

#[test]
fn test_test_server() {
    let server = TestServer::new().unwrap();

    let mut resource_404 = server.create_resource("/");
    resource_404
        .status(http_test_server::http::Status::NotFound)
        .method(http_test_server::http::Method::GET)
        .body(r"404 NOT HELLO");

    let mut resource2 = server.create_resource("/hello");
    resource2
        .status(http_test_server::http::Status::Accepted)
        .method(http_test_server::http::Method::GET)
        .header("Content-Type", "application/json")
        .header("Cache-Control", "no-cache")
        .body(r"Hello!");

    let client = reqwest::blocking::Client::new();
    let link = format!("http://localhost:{}/hello", server.port());
    let request = client.get(&link);
    match request.send() {
        Err(err) => { println!("ERR: {:?}", err); },
        Ok(response) => {
            let response_str = response.text().unwrap();
            assert_eq!(response_str, r"Hello!");
        }
    };
}
```
and this will return 
```
running 1 test
thread 'myapp::test::test_test_server' panicked at 'assertion failed: `(left == right)`
  left: `"404 NOT HELLO"`,
 right: `"Hello!"`', src/myapp/test.rs:1834:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test myapp::test::test_test_server ... FAILED

failures:

failures:
    myapp::test::test_test_server

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 169 filtered out; finished in 0.01s
```
It is my understanding that `create_resource` simply defines one additional entry in a list of resources, so why this doesn't give the expected result somewhat baffles me. Is this a bug or am I doing something wrong?

Thanks,
Marc

EDIT:  An _internal_ test might look something like this inside your test suite:
```
    #[test]
    fn should_handle_multiple_resources() {
        let server = TestServer::new().unwrap();
        let resource = server.create_resource("/something");
        resource.status(Status::OK).body("<some body>");

        let resource2 = server.create_resource("/something-else");
        resource2.status(Status::OK).body("<some other body>");

        assert_eq!(resource.request_count(), 0);

        let _ = make_request(server.port(), "/something");
        let _ = make_request(server.port(), "/something-else");

        thread::sleep(Duration::from_millis(200));

        assert_eq!(resource.request_count(), 1);
        assert_eq!(resource2.request_count(), 1);
    }
```

The above returns
```
failures:

---- tests::should_handle_multiple_resources stdout ----
thread 'tests::should_handle_multiple_resources' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`', src/lib.rs:670:9
```

EDIT 2:
It turns out the regex is slightly permissive, a resource "/some" will also match "/something-longer".
This is easily worked around by enforcing a trailing slash to each resource.

The main issue. however, appears to be in `find_resource`, see FIXME comment:
```
fn find_resource(method: String, url: String, resources: ServerResources) -> Resource {
    let resources = resources.lock().unwrap();
    let resources_for_uri = resources.iter().filter(|r| r.matches_uri(&url));

    if resources_for_uri.count() == 0 {
        return Resource::new(&url).status(Status::NotFound).clone();
    }

    // FIXME: The next line should probably not iterate `resources` but `resources_for_uri`
    match resources.iter().find(|r| { r.get_method().equal(&method) }) {
        Some(resource) => {
            resource.increment_request_count();
            resource.clone()
        },
        None => Resource::new(&url).status(Status::MethodNotAllowed).clone()
    }
}
```

Without diving deeper it seems slightly tricky to keep Rust happy about running `find` on a `filter` result, so it makes sense to immediately iterate both URI and request method, then do more filtering in the rare case an error occurs to distinguish between 404 or method not allowed:
```
fn find_resource(method: String, url: String, resources: ServerResources) -> Resource {
    let resources = resources.lock().unwrap();

    match resources.iter().find(|r| r.matches_uri(&url) && r.get_method().equal(&method) ) {
        Some(resource) => {
            resource.increment_request_count();
            resource.clone()
        },
        None => {
            let resources_for_uri = resources.iter().filter(|r| r.matches_uri(&url));
            if resources_for_uri.count() == 0 {
                return Resource::new(&url).status(Status::NotFound).clone();
            }
            Resource::new(&url).status(Status::MethodNotAllowed).clone()
        }
    }
}
```
For the typical test scenario, this will hit the correct (existing) resource right away.
The amount of resources to iterate tends to be very small (up until now, just the one!), so it would be expected that the performance impact of (re)filtering in case of errors would be minimal.

**Comment by viniciusgerevini** [comment]
Hello Marc. Sorry the late response. Thanks a lot for looking into this and for the detailed issue report. That was a big oversight.

I agree with your assessment. stopping as soon the right match is found is smart.

Just to complement your detailed investigation, in the scenario where you are saying that you need to work on the iterator after doing a find,  we could have solved it by doing like this:
``` rust
   let resources_for_uri = resources.iter().filter(|r| r.matches_uri(&url));

   // this one clones the reference before count
    if resources_for_uri.clone().count() == 0 {
        return Resource::new(&url).status(Status::NotFound).clone();
    }
   // into_iter transforms the filter back into an interator
    match resources_for_uri.into_iter().find(|r| { r.get_method().equal(&method) }) {
        Some(resource) => {
            resource.increment_request_count();
            resource.clone()
        },
        None => Resource::new(&url).status(Status::MethodNotAllowed).clone()
    }

```
That's a smaller change, but  as you pointed out it's more likely to have a match than a no match, so doing the find first is a better option.

Thanks again. I'll merge your changes now. I don´t remember if the pipeline is deploying automatically, but if not I'll be releasing it today or tomorrow. Cheers.

**Comment by viniciusgerevini** [comment]
All done. Shipped version 2.1.1. :rocket:
